gh-142352: Fix start_tls() to transfer buffered data from StreamReader

kasimov-maxim opened the pull request:
Fixes #142352

## Summary

When using `StreamWriter.start_tls()` to upgrade a connection to TLS mid-stream, any data already buffered in the `StreamReader` was lost. This commonly occurs when implementing servers that support the PROXY protocol.

## Changes

- Added `_transfer_buffered_data_to_ssl()` method to `BaseEventLoop` that transfers buffered data from `StreamReader._buffer` to `SSLProtocol._incoming` before the TLS handshake begins
- Added tests for the PROXY protocol + TLS scenario

## Generative AI usage
Some parts of this PR were assisted by a generative AI tool. Specifically:

- help with drafting English-language comments and documentation
- suggestions for test refactoring
- assistance with wording and structuring the PR description

All functional logic and final edits were reviewed, verified, and authored by
me, in accordance with the Python devguide guidelines.


## Test plan

- [x] Added new tests in `test_streams.py`
- [x] All asyncio tests pass
- [x] `make patchcheck` passes

<!-- gh-issue-number: gh-142352 -->
* Issue: gh-142352
<!-- /gh-issue-number -->

kasimov-maxim committed:
gh-142352: asyncio.streams: transfer buffered data to SSL layer in start_tls()

kumaraditya303 commented on Lib/asyncio/base_events.py:
```diff
@@ -1311,6 +1311,30 @@ async def _sendfile_fallback(self, transp, file, offset, count):
                 file.seek(offset + total_sent)
             await proto.restore()
 
+    def _transfer_buffered_data_to_ssl(self, protocol, ssl_protocol):
```
As this method is called from just one place, I suggest to inline it

kasimov-maxim replied:
I kept this logic in a small helper because it represents a distinct
conceptual step (bridging StreamReader buffering into the SSLProtocol BIO),
and `start_tls()` is already quite dense. Even though the block is relatively
short, it introduces non-obvious buffering semantics involving private
StreamReader and SSLProtocol internals.

Keeping it named helps document the behavior required for mid-connection
`start_tls()` scenarios, avoids cluttering the main flow, and leaves room
for future adjustments if this logic needs to grow.

If we inline this, I think it would be worth keeping a fairly detailed
comment explaining why this buffering transfer is needed.

That said, I'm happy to inline it if you think that's preferable here.

picnixz replied:
There is no need for a method as it would indirectly be exposed on subclasses. The docstring is also un-necessary, and there is no need to mention custom implementations.

> If we inline this, I think it would be worth keeping a fairly detailed
comment explaining why this buffering transfer is needed.

No. Just link an issue (and possibly a short comment). The code base prefers concise comments. It's not meant to be read by everyone and we definitely don't want to have more docstrings than code. If something is meant to be extensively documented it should be done at the RST level.

kasimov-maxim replied:
Understood, thanks. I'll inline it and keep only a short comment with an issue reference.

webknjaz replied:
*(my $0.02)*

> no need for a method as it would indirectly be exposed on subclasses

FWIW, I often make use of the technique of naming things. It's nice to have clear semantic names for such abstractions: it'd be more difficult to make sense of `ssl_protocol._incoming.write(protocol._stream_reader._buffer)` in the context of `start_tls()`. Plus having these low-level details leak into that layer would mean that future refactorings may result in unrelated lines being added in-between, making their way into a block of otherwise tightly-coupled instructions.

Clearly, having it a member of the class is suboptimal but what's wrong with having a `_private` helper function around, for example?

picnixz replied:
We do not like proliferation of methods when code can be inlined. Making sense of  line is the reason why adding a small comment is enough.

We also should not be worried about the future. However I would recommend having a more compact block (we avoid blank lines in the stdlib and only use them to separate sections in a function). 

Now, If we want to make it private, why not also. But I do not think that it is better to leak the method itself to subclasses (because then we need to indicate that the logic is private and neither protected or public; this the unfortunate part of not having a protected convention aside from `@override`).

Since Kumar is the code owner I think we should respect his intentions on that matter. Though they could have changed after your suggestion.

picnixz commented on Lib/test/test_asyncio/test_streams.py:
```diff
@@ -819,6 +819,165 @@ async def client(addr):
         self.assertEqual(msg2, b"hello world 2!\n")
 
+    def _run_test_start_tls_behind_proxy(self, send_combined):
+        """Test start_tls() when TLS ClientHello arrives with PROXY header.
```
Remove docstrings from tests. It would be shown on failure. Use regular comments and shorten it. We don't want to restate the entire issue so simply link the issue's number as well.

kasimov-maxim replied:
Understood, will update the tests accordingly.

picnixz commented on Lib/test/test_asyncio/test_streams.py:
```diff
@@ -819,6 +819,165 @@ async def client(addr):
+        test_message = b"hello world\n"
+        expected_response = reverse_message(test_message)
+
+        class TCPProxyServer:
```
Do we really need this class? don't we already have some helpers elsewhere? The test also looks overly complex. Why not just patching existing TCPServer? (if possible; I don't know if there isn't some class that could be used as is with some minimal modifications)

picnixz reviewed (dismissed):
Has this PR been generated using an LLM? if so, indicate which places were so and read https://devguide.python.org/getting-started/generative-ai/.

webknjaz commented on Lib/asyncio/base_events.py:
```diff
@@ -1341,6 +1365,8 @@ async def start_tls(self, transport, protocol, sslcontext, *,
+
         # Pause early so that "ssl_protocol.data_received()" doesn't
         # have a chance to get called before "ssl_protocol.connection_made()".
         transport.pause_reading()
```
Wouldn't it make sense to transfer the buffer after reading is paused rather than before?

kasimov-maxim replied:
Moving the buffer is memory-only work without `await`, so doing it before or after `pause_reading()` is practically equivalent, as long as it happens before `transport.set_protocol()` and before `connection_made()/resume_reading()`. So the current order is safe.

webknjaz replied:
I thought, it might make sense semantically. As a casual reader, seeing a comment after the data manipulation, that says we're pausing before undesired data reads happen seemed confusing to me. Perhaps, others won't feel the same, though.

kumaraditya303 replied:
I agree it would be clearer to do this after transport is paused even if both are same.

kasimov-maxim replied:
@kumaraditya303 

OK. Done — paused before the buffer handoff for clarity.

webknjaz commented on Lib/asyncio/base_events.py:
```diff
@@ -1311,6 +1311,30 @@ async def _sendfile_fallback(self, transp, file, offset, count):
+        buffer = stream_reader._buffer
+        if buffer:
+            ssl_protocol._incoming.write(buffer)
+            buffer.clear()
```
I wonder if this could better belong @ https://github.com/python/cpython/blob/b2ed3c3c26ad8163bc1f56824f4b1db4fc75af75/Lib/asyncio/sslproto.py#L322.

Technically, there could be something like

```diff
-         self._incoming = ssl.MemoryBIO()
+         self._incoming = ssl.MemoryBIO()
+         try:
+             underlying_raw_stream_buffer = app_protocol._stream_reader._buffer
+         except AttributeError:
+             pass
+         else:
+             self._incoming.write(underlying_raw_stream_buffer)
+             underlying_raw_stream_buffer.clear()
```

But I don't know how other places that initialize `sslproto.SSLProtocol()` would be affected. So it's just an idea to check out and see if other instances should have the same buffer transfer logic (looks like it's used in two other modules).

kasimov-maxim committed:
gh-142352: simplify start_tls buffered-data test

Replace the proxy-based combined/separate cases with a deterministic
buffered-ClientHello check and a ping/pong verification after start_tls.

kasimov-maxim committed:
gh-142352: pause transport before moving buffered data

Move pause_reading() ahead of the StreamReader buffer handoff in
BaseEventLoop.start_tls() for clearer sequencing.

fantix commented on Lib/asyncio/base_events.py:
```diff
@@ -1345,6 +1345,15 @@ async def start_tls(self, transport, protocol, sslcontext, *,
 
+        # gh-142352: move buffered StreamReader data to SSLProtocol
+        if server_side:
+            stream_reader = getattr(protocol, '_stream_reader', None)
```
Maybe we should check the specific `StreamReaderProtocol` type? Just in case user protocols happened to have the same but totally irrelevant `_stream_reader` property.

kasimov-maxim replied:
If the concern is the semantics of `_stream_reader`, we can instead gate on the reader itself: either check that it’s a `StreamReader`, or at least that it exposes a `_buffer` of type `bytearray`. This keeps the fix focused on the `StreamReader` contract without hard‑coding a specific protocol class.

`loop.start_tls()` is a low‑level API and doesn’t know the semantics of user protocol buffers. So forwarding should only happen when we have a clear, explicit contract (i.e. `StreamReader`), not for arbitrary protocols.

fantix replied:
Thanks for the reply! Though, I'm not sure I fully understand you (like why hard-coding `StreamReader` is okay while `StreamReaderProtocol` is not? Isn't type-checking `StreamReaderProtocol` more on the side of "not for arbitrary protocols"? But you don't have to answer these, I'm just explaining where I failed to understand 🥲 sorry!). Please allow me to reiterate on my idea here:

I just felt weird to me that you chose this contract (3) out of the 4 options as follows, where the least StreamReader-specific goes first:

1. Explicit parameter like `loop.start_tls(forward_handshake_data=...)`
2. New internal contract on the protocol like `protocol._forward_tls_handshake_data`
3. Status quo contract like `protocol. _stream_reader._buffer`
4. Type checking like `isinstance(protocol, StreamReaderProtocol)`

My theory is like, if we cannot choose (1) for backport reasons, we should pick from the other end of the list, which is (4), as a patch that only addresses the `StreamWriter.start_tls()` issue alone, and introduces nothing more, no hidden features. (2) and (3) are just workarounds of the backport limit (of no new params) to me; they are ultimately equivalent to adding a new parameter, other than the stability of the API. For example, one can do this:

```python
class MockStreamReader:
    @property
    def _buffer(self):
        ....

class MyProtocol:
    @property
    def _stream_reader(self):
        return MockStreamReader(self)
```

And achieve the same thing as (1). If I have to, I prefer (2) over (3) for API explicitness, even though it is meant to be internal. But again, I'd say (1) if possible, or else (4), then (2) and (3).

kasimov-maxim replied:
> in case _user protocols_ happened to have the same but totally irrelevant `_stream_reader` property

In your first comment you effectively acknowledge that protocols other than `StreamReaderProtocol` might exist and could expose an `_stream_reader` with different semantics, which is exactly why a bare `getattr(protocol, '_stream_reader', None)` is too weak.

My suggestion was to **gate on the reader itself**, not the protocol class. Concretely:

```py
stream_reader = getattr(protocol, '_stream_reader', None)
if isinstance(stream_reader, StreamReader):
    ...
```

That doesn’t hard‑code `StreamReaderProtocol`, but it *does* verify the semantics we actually rely on (a real `StreamReader` with the standard buffering contract). It also leaves room for the kind of non‑`StreamReaderProtocol` protocols you mentioned, as long as they genuinely use a `StreamReader`.

So to summarize the trade‑off:

- `isinstance(protocol, StreamReaderProtocol)` is the strictest and safest, but overly narrow.
- `isinstance(stream_reader, StreamReader)` targets the actual buffer contract we need, without binding to one protocol class.
- A bare `stream_reader = getattr(protocol, '_stream_reader', None)` is too permissive.

Happy to switch to (4) if the priority is minimal backportable change, but I think checking the reader type is a reasonable middle ground.

fantix replied:
Thank you for the explanation! That's very helpful.

(I think now I understand:

1. Your PR targets (3)
2. I proposed (4) with too little context and confused you with the semantics of `_stream_reader`
3. You think (4) is too narrow (?) and proposed (3.5) `isinstance(stream_reader, StreamReader)`
4. I got confused in turn and explained the full (1) - (4)
5. You explained everything else ❤️ 
)

So my question now is, why do you think (4) is overly narrow? Is there a particular use case that you are trying to support, like a custom protocol class using `StreamReader` internally? If so, I believe the proper way should be either (1) for future versions, or the `BufferedProtocol` workaround I mentioned earlier, instead of relying on unstable internal contracts like (2), (3) or (3.5).

> Happy to switch to (4) if the priority is minimal backportable change, but I think checking the reader type is a reasonable middle ground.

Ohh that would be something for @kumaraditya303 to decide, I don't have strong opinions per se. Thank you for bouncing ideas with me!

kumaraditya303 replied:
I am thinking of backporting this to 3.13 and 3.14 and for that I would prefer the the minimal version which is to implement @fantix's suggestion of (4). We can do a more complete fix for the main branch in a followup.

kumaraditya303 commented on Lib/test/test_asyncio/test_streams.py:
```diff
@@ -819,6 +819,51 @@ async def client(addr):
+
+        async def server_handler(client_reader, client_writer):
+            # Wait for TLS ClientHello to be buffered before start_tls().
+            await asyncio.wait_for(
```
I might be missing something but I don't think there is any reason to wrap these in `asyncio.wait_for`.

kumaraditya303 committed:
address review

kumaraditya303 committed:
fix import

fantix reviewed (commented):
LGTM!
